README.rst
==============
``simpleldap``
==============


About
=====

A small wrapper around the python-ldap library that provides a more Pythonic
interface for LDAP server connections, LDAP objects, and the common get and
search operations.


Installation
============

Install using pip::

    pip install simpleldap


Examples
========

A quick and easy example using ``simpleldap``::

    >>> import simpleldap
    >>> conn = simpleldap.Connection('directory.example.com')
    >>> user = conn.get("uid=myuser")
    >>> user.dn
    'uid=myuser,dc=directory,dc=example,dc=com'
    >>> user['cn']
    ['Joe Smith', 'Joe M. Smith']
    >>> user.first('cn')
    'Joe Smith'
    >>> conn.close()

The ``Connection`` object can also be used as a context manager, e.g.::

    with simpleldap.Connection('directory.example.com') as conn:
        users = conn.search("(&(givenName=Joe)(sn=Smith))")

A common method for authenticating users is to connect to an LDAP server using
a service user/account and then attempt a bind operation using the user's
credentials (i.e. DN and password).  The ``authenticate`` method makes this
simple::

    with simpleldap.Connection('directory.example.com') as conn:
        is_valid = conn.authenticate('uid=myuser,dc=directory,dc=example,dc=com', 'password')

.. note::
    The ``authentication`` method does not perform an unbind and does not bind again
    using the original connection's credentials; therefore, any further
    actions following a successful ``authenticate`` call will be performed as
    the authenticated user.

LDAP also offers a feature to compare an attribute's value with a given string.
This can occasionally be more efficient and expressive than grabbing an entire
object from the LDAP store. ``simpleldap`` offers a ``compare`` method for this
feature::

    >>> conn = simpleldap.Connection('directory.example.com')
    >>> user_dn = 'uid=myuser,dc=directory,dc=example,dc=com'
    >>> conn.compare(user_dn, 'cn', 'Joe Smith')
    True



CHANGES.rst
=========
Changelog
=========


0.8 - May 28, 2014
==================
* Dropped support for Python 2.5.
* Removed distribute_setup bootstrap file, installation now requires
  setuptools to already be installed.
* #1: Added an ``authenticate`` method for authenticating a given DN and
  password using a bind operation.
* #2: Added the ability to set (and clear) defaults for the ``search`` method.
* #4: Added a ``compare`` method, wrapping ldap's compare, to return boolean
  values instead of 0 and 1.


0.7.1 - July 19, 2012
=====================
* Added tox testing configuration, setup to run against Python 2.5, 2.6,
  and 2.7.
* Converted repo from hg to git.


0.7 - July 10, 2011
===================
* Initial release.
* ``Connection`` class for conveniently creating connections to LDAP servers.
* ``LDAPItem`` class for wrapping LDAP result objects and allowing more
  convenient, dictionary-like access to their contained data.


tests/tests.py
from __future__ import with_statement  # Python 2.5 compatibility.
from operator import itemgetter
from unittest import TestCase

import ldap

import simpleldap


class ConnectionTests(TestCase):

    hosts = (
        # host, port, encryption type, require_cert.
        ('ldap.utexas.edu', None, None, None),
        ('ldap.utexas.edu', 389, None, None),
        ('ldap.ucdavis.edu', None, 'tls', False),
        ('ldap.ucdavis.edu', 389, 'tls', False),
        ('ldap.ucdavis.edu', None, 'ssl', False),
        ('ldap.ucdavis.edu', 636, 'ssl', False),
    )

    def test_connect(self):
        for host, port, method, cert in self.hosts:
            try:
                conn = simpleldap.Connection(hostname=host, port=port,
                                             encryption=method,
                                             require_cert=cert)
            except Exception, e:
                self.fail("Got error connecting to %s %s %s %s: %s"
                          % (host, port, method, cert, e))
            else:
                conn.close()

    def test_initialize_kwargs(self):
        from StringIO import StringIO
        output = StringIO()
        initialize_kwargs = {'trace_file': output, 'trace_level': 0}
        conn = simpleldap.Connection('ldap.utexas.edu',
                                     initialize_kwargs=initialize_kwargs)
        conn.close()
        self.assertFalse(output.getvalue())
        initialize_kwargs = {'trace_file': output, 'trace_level': 1}
        conn = simpleldap.Connection('ldap.utexas.edu',
                                     initialize_kwargs=initialize_kwargs)
        conn.close()
        self.assertTrue(output.getvalue())

    def test_context_manager(self):
        host, port, method, cert = self.hosts[0]
        with simpleldap.Connection(hostname=host, port=port, encryption=method,
                                   require_cert=cert) as conn:
            conn.connection.whoami_s()

    def test_invalid_encryption(self):
        self.assertRaises(simpleldap.InvalidEncryptionProtocol,
                          simpleldap.Connection,
                          hostname='ldap.ucdavis.edu', encryption='foo')

    def test_connection_options(self):
        opt = 'OPT_TIMELIMIT'
        value = 1000
        conn = simpleldap.Connection(hostname='ldap.utexas.edu',
                                     options={opt: value},
                                     # No way to really test debug output, but
                                     # thrown in for coverage.
                                     debug=True)
        self.assertEqual(conn.connection.get_option(getattr(ldap, opt)), value)

    def test_search(self):
        conn = simpleldap.Connection('ldap.ucdavis.edu')
        objs = conn.search('cn=*', base_dn='ou=Groups,dc=ucdavis,dc=edu')
        self.assertTrue(len(objs) > 3)
        for obj in objs:
            self.assertTrue(isinstance(obj, conn.result_item_class))

    def test_search_params(self):
        conn = simpleldap.Connection('ldap.ucdavis.edu')
        self.assertRaises(ldap.SIZELIMIT_EXCEEDED, conn.search, 'cn=*',
                          base_dn='ou=Groups,dc=ucdavis,dc=edu', limit=1)
        kwargs = {'filter': 'cn=External Anonymous',
                  'base_dn': 'ou=Groups,dc=ucdavis,dc=edu'}
        # Should return all attrs.
        self.assertTrue(len(conn.search(**kwargs)[0]) > 2)
        # Should return just cn attr.
        obj = conn.search(attrs=['cn'], **kwargs)[0]
        self.assertEqual(len(obj), 1)
        self.assertTrue('cn' in obj)

    def test_search_defaults(self):
        conn = simpleldap.Connection('ldap.ucdavis.edu', search_defaults={'limit': 1})
        conn.set_search_defaults(base_dn='ou=Groups,dc=ucdavis,dc=edu')
        self.assertRaises(ldap.SIZELIMIT_EXCEEDED, conn.search, 'cn=*')
        kwargs = {'filter': 'cn=External Anonymous', }
        conn.clear_search_defaults(['limit'])
        # Should return all attrs.
        self.assertTrue(len(conn.search(**kwargs)[0]) > 2)
        # Should return just cn attr.
        conn.set_search_defaults(attrs=['cn'])
        obj = conn.search(**kwargs)[0]
        self.assertEqual(len(obj), 1)
        self.assertTrue('cn' in obj)
        conn.clear_search_defaults()
        self.assertEqual(conn._search_defaults, {})

    def test_get(self):
        conn = simpleldap.Connection('ldap.ucdavis.edu')
        obj = conn.get('cn=External Anonymous',
                       base_dn='ou=Groups,dc=ucdavis,dc=edu')
        self.assertTrue(isinstance(obj, conn.result_item_class))
        self.assertEqual(obj['cn'], ['External Anonymous'])

        self.assertRaises(simpleldap.ObjectNotFound, conn.get,
                          'cn=Does not exist',
                          base_dn='ou=Groups,dc=ucdavis,dc=edu')
        self.assertRaises(simpleldap.MultipleObjectsFound, conn.get, 'cn=*',
                          base_dn='ou=Groups,dc=ucdavis,dc=edu')

    def test_compare(self):
        conn = simpleldap.Connection('ldap.ucdavis.edu')
        obj = conn.get('cn=External Anonymous',
                       base_dn='ou=Groups,dc=ucdavis,dc=edu')
        self.assertTrue(conn.compare(obj.dn, 'cn', 'External Anonymous'))
        self.assertFalse(conn.compare(obj.dn, 'cn', 'foo'))


class AuthenticateTests(TestCase):

    def test_success(self):
        conn = simpleldap.Connection('ldap.ucdavis.edu')
        self.assertTrue(conn.authenticate('cn=External Anonymous,ou=Groups,dc=ucdavis,dc=edu', ''))

    def test_fail_no_such_object(self):
        conn = simpleldap.Connection('ldap.ucdavis.edu')
        self.assertFalse(conn.authenticate('uid=foobar', 'baz'))

    def test_fail_unwilling_to_perform(self):
        conn = simpleldap.Connection('ldap.utexas.edu')
        self.assertFalse(conn.authenticate('cn=Anonymous', ''))

    def test_fail_invalid_credentials(self):
        conn = simpleldap.Connection('ldap.utexas.edu')
        self.assertFalse(conn.authenticate('uid=foobar', 'baz'))


class LDAPItemTests(TestCase):

    mock_results = [
        ('uid=joe,ou=people,dc=directory,dc=example,dc=com',
            {'cn': ['Joseph Smith', 'Joseph Smith Jr.'],
             'givenName': ['Joseph'],
             'sn': ['Smith'],
             'displayName': ['Joe L. Smith', 'Joe Smith'],
            }),
        ('uid=bob,ou=people,dc=directory,dc=example,dc=com',
            {'cn': ['Robert West', 'Robert T. West'],
             'givenName': ['Robert'],
             'sn': ['West'],
             'displayName': ['Bob West'],
            }),
    ]

    def test_dn(self):
        item = simpleldap.LDAPItem(self.mock_results[0])
        self.assertEqual(item.dn,
                         'uid=joe,ou=people,dc=directory,dc=example,dc=com')

    def test_str(self):
        item = simpleldap.LDAPItem(self.mock_results[0])
        self.assertEqual(str(item), """\
         cn: Joseph Smith
             Joseph Smith Jr.
displayName: Joe L. Smith
             Joe Smith
  givenName: Joseph
         sn: Smith""")

    def test_item_lookup(self):
        item = simpleldap.LDAPItem(self.mock_results[0])
        self.assertEqual(item['cn'], ['Joseph Smith', 'Joseph Smith Jr.'])
        self.assertRaises(KeyError, itemgetter('foo'), item)

    def test_first_lookup(self):
        item = simpleldap.LDAPItem(self.mock_results[0])
        self.assertEqual(item.first('cn'), 'Joseph Smith')
        self.assertEqual(item.first('sn'), 'Smith')
        self.assertRaises(KeyError, item.first, 'foo')

    def test_case_insensitive(self):
        item = simpleldap.LDAPItem(self.mock_results[0])
        self.assertEqual(item['cn'], ['Joseph Smith', 'Joseph Smith Jr.'])
        self.assertEqual(item['cn'], item['CN'])
        self.assertEqual(item.first('cn'), item.first('CN'))
        self.assertTrue('cn' in item)
        self.assertTrue('CN' in item)

    def test_value_contains(self):
        item = simpleldap.LDAPItem(self.mock_results[0])
        self.assertTrue(item.value_contains('Joseph', 'cn'))
        self.assertFalse(item.value_contains('Bob', 'cn'))

    def test_equals(self):
        self.assertEqual(simpleldap.LDAPItem(self.mock_results[0]),
                         simpleldap.LDAPItem(self.mock_results[0]))
        self.assertNotEqual(simpleldap.LDAPItem(self.mock_results[0]),
                            simpleldap.LDAPItem(self.mock_results[1]))


if __name__ == "__main__":
    import unittest
    unittest.main()


tests/cidict.py
from unittest import TestCase

from simpleldap.cidict import cidict


class cidictTests(TestCase):

    def test_setitem_getitem_get(self):
        d = cidict()
        d['foo'] = 'bar'
        self.assertEqual(d['foo'], 'bar')
        self.assertEqual(d['FOO'], 'bar')
        self.assertEqual(d.get('foo'), 'bar')
        self.assertEqual(d.get('FOO'), 'bar')
        d['FOO'] = 'baz'
        self.assertEqual(d['foo'], 'baz')
        self.assertEqual(d['FOO'], 'baz')
        self.assertEqual(d.get('foo'), 'baz')
        self.assertEqual(d.get('FOO'), 'baz')

    def test_update(self):
        d = cidict({'foo': 'bar', 'FOO': 'bar', 'bar': 'baz'})
        self.assertEqual(len(d), 2)

    def test_keys_has_key(self):
        d = cidict({'foo': 'bar', 'bar': 'baz'})
        self.assertEqual(sorted(list(d)), ['bar', 'foo'])
        self.assertEqual(sorted(d.keys()), ['bar', 'foo'])
        self.assertTrue(d.has_key('foo'))
        self.assertTrue(d.has_key('FOO'))
        self.assertTrue('foo' in d)
        self.assertTrue('FOO' in d)
        d['FOO'] = 'bar'
        del d['foo']
        self.assertFalse(d.has_key('foo'))
        self.assertFalse(d.has_key('FOO'))
        self.assertFalse('foo' in d)
        self.assertFalse('FOO' in d)

    def test_values_items(self):
        d = cidict({'foo': 'bar', 'FOO': 'bar', 'bar': 'baz'})
        self.assertEqual(sorted(d.values()), ['bar', 'baz'])
        d = cidict({'foo': 'bar', 'bar': 'baz'})
        self.assertEqual(sorted(d.items()), [('bar', 'baz'), ('foo', 'bar')])

if __name__ == "__main__":
    import unittest
    unittest.main()


simpleldap/cidict.py
"""
A re-implementation of the ldap module's cidict that inherits from dict instead
of UserDict so that we can, in turn, inherit from cidict.
"""


class cidict(dict):
    """
    Case-insensitive but case-respecting dictionary.
    """

    def __init__(self, default=None):
        self._keys = {}
        super(cidict, self).__init__({})
        self.update(default or {})

    def __getitem__(self, key):
        return super(cidict, self).__getitem__(key.lower())

    def __setitem__(self, key, value):
        lower_key = key.lower()
        self._keys[lower_key] = key
        super(cidict, self).__setitem__(lower_key, value)

    def __delitem__(self, key):
        lower_key = key.lower()
        del self._keys[lower_key]
        super(cidict, self).__delitem__(lower_key)

    def update(self, dict):
        for key in dict:
            self[key] = dict[key]

    def has_key(self, key):
        return super(cidict, self).has_key(key.lower())

    __contains__ = has_key

    def get(self, key, *args, **kwargs):
        return super(cidict, self).get(key.lower(), *args, **kwargs)

    def keys(self):
        return self._keys.values()

    def items(self):
        return [(k, self[k]) for k in self.keys()]


simpleldap/__init__.py
"""
This module makes simple LDAP queries simple.
"""

import ldap

from simpleldap.cidict import cidict


#
# Exceptions.
#

class SimpleLDAPException(Exception):
    """Base class for all simpleldap exceptions."""


class ObjectNotFound(SimpleLDAPException):
    """
    Exception when no objects were returned, but was expecting a single item.
    """


class MultipleObjectsFound(SimpleLDAPException):
    """
    Exception for when multiple objects were returned, but was expecting only
    a single item.
    """


class ConnectionException(Exception):
    """Base class for all Connection object exceptions."""


class InvalidEncryptionProtocol(ConnectionException):
    """Exception when given an unsupported encryption protocol."""


#
# Classes.
#

class LDAPItem(cidict):
    """
    A convenience class for wrapping standard LDAPResult objects.
    """

    def __init__(self, result):
        super(LDAPItem, self).__init__()
        self.dn, self.attributes = result
        # XXX: quick and dirty, should really proxy straight to the existing
        # self.attributes dict.
        for attribute, values in self.attributes.iteritems():
            # Make the entire list of values for each LDAP attribute
            # accessible through a dictionary mapping.
            self[attribute] = values

    def first(self, attribute):
        """
        Return the first value for the given LDAP attribute.
        """
        return self[attribute][0]

    def value_contains(self, value, attribute):
        """
        Determine if any of the items in the value list for the given
        attribute contain value.
        """
        for item in self[attribute]:
            if value in item:
                return True
        return False

    def __str__(self):
        """
        Print attribute names and values, one per line, in alphabetical order.

        Attribute names are displayed right-aligned to the length of the
        longest attribute name.
        """
        attributes = self.keys()
        longestKeyLength = max([len(attr) for attr in attributes])
        output = []
        for attr in sorted(attributes):
            values = ("\n%*s  " % (longestKeyLength, ' ')).join(self[attr])
            output.append("%*s: %s" % (longestKeyLength, attr, values))
        return "\n".join(output)

    def __eq__(self, other):
        return self.dn == other.dn


class Connection(object):
    """
    A connection to an LDAP server.
    """

    # The class to use for items returned in results.  Subclasses can change
    # this to a class of their liking.
    result_item_class = LDAPItem

    # List of exceptions to treat as a failed bind operation in the
    # authenticate method.
    failed_authentication_exceptions = [
        ldap.NO_SUCH_OBJECT,  # e.g. dn matches no objects.
        ldap.UNWILLING_TO_PERFORM,  # e.g. dn with no password.
        ldap.INVALID_CREDENTIALS,  # e.g. wrong password.
    ]

    def __init__(self, hostname='localhost', port=None, dn='', password='',
                 encryption=None, require_cert=None, debug=False,
                 initialize_kwargs=None, options=None, search_defaults=None):
        """
        Bind to hostname:port using the passed distinguished name (DN), as
        ``dn``, and password.

        If ``hostname`` is not given, default to ``'localhost'``.

        If no user and password is given, try to connect anonymously with a
        blank DN and password.

        ``encryption`` should be one of ``'tls'``, ``'ssl'``, or ``None``.
        If ``'tls'``, then the standard port 389 is used by default and after
        binding, tls is started.  If ``'ssl'``, then port 636 is used by
        default.  ``port`` can optionally be given for connecting to a
        non-default port.

        ``require_cert`` is None by default.  Set this to ``True`` or
        ``False`` to set the ``OPT_X_TLS_REQUIRE_CERT`` ldap option.

        If ``debug`` is ``True``, debug options are turned on within ldap and
        statements are ouput to standard error.  Default is ``False``.

        If given, ``options`` should be a dictionary of any additional
        connection-specific ldap  options to set, e.g.:
        ``{'OPT_TIMELIMIT': 3}``.

        If given, ``search_defaults`` should be a dictionary of default
        parameters to be passed to the search method.
        """
        if search_defaults is None:
            self._search_defaults = {}
        else:
            self._search_defaults = search_defaults

        if not encryption or encryption == 'tls':
            protocol = 'ldap'
            if not port:
                port = 389
        elif encryption == 'ssl':
            protocol = 'ldaps'
            if not port:
                port = 636
        else:
            raise InvalidEncryptionProtocol(
                "Invalid encryption protocol, must be one of: 'tls' or 'ssl'.")

        if require_cert is not None:
            ldap.set_option(ldap.OPT_X_TLS_REQUIRE_CERT, require_cert)
        if debug:
            ldap.set_option(ldap.OPT_DEBUG_LEVEL, 255)
        else:
            ldap.set_option(ldap.OPT_DEBUG_LEVEL, 0)

        uri = '%s://%s:%s' % (protocol, hostname, port)
        if initialize_kwargs:
            self.connection = ldap.initialize(uri, **initialize_kwargs)
        else:
            self.connection = ldap.initialize(uri)
        if options:
            for name, value in options.iteritems():
                self.connection.set_option(getattr(ldap, name), value)
        if encryption == 'tls':
            self.connection.start_tls_s()
        self.connection.simple_bind_s(dn, password)

    def set_search_defaults(self, **kwargs):
        """
        Set defaults for search.

        Examples::

            conn.set_search_defaults(basedn='dc=example,dc=com', timeout=100)
            conn.set_search_defaults(attrs=['cn'], scope=ldap.SCOPE_BASE)
        """
        self._search_defaults.update(kwargs)

    def clear_search_defaults(self, args=None):
        """
        Clear all search defaults specified by the list of parameter names
        given as ``args``.  If ``args`` is not given, then clear all existing
        search defaults.

        Examples::

            conn.set_search_defaults(scope=ldap.SCOPE_BASE, attrs=['cn'])
            conn.clear_search_defaults(['scope'])
            conn.clear_search_defaults()
        """
        if args is None:
            self._search_defaults.clear()
        else:
            for arg in args:
                if arg in self._search_defaults:
                    del self._search_defaults[arg]

    def __enter__(self):
        return self

    def __exit__(self, exc_type, exc_value, traceback):
        self.close()

    def close(self):
        """
        Shutdown the connection.
        """
        self.connection.unbind_s()

    def search(self, filter, base_dn=None, attrs=None, scope=None,
               timeout=None, limit=None):
        """
        Search the directory.
        """
        if base_dn is None:
            base_dn = self._search_defaults.get('base_dn', '')
        if attrs is None:
            attrs = self._search_defaults.get('attrs', None)
        if scope is None:
            scope = self._search_defaults.get('scope', ldap.SCOPE_SUBTREE)
        if timeout is None:
            timeout = self._search_defaults.get('timeout', -1)
        if limit is None:
            limit = self._search_defaults.get('limit', 0)

        results = self.connection.search_ext_s(
            base_dn, scope, filter, attrs, timeout=timeout, sizelimit=limit)
        return self.to_items(results)

    def get(self, *args, **kwargs):
        """
        Get a single object.

        This is a convenience wrapper for the search method that checks that
        only one object was returned, and returns that single object instead
        of a list.  This method takes the exact same arguments as search.
        """
        results = self.search(*args, **kwargs)
        num_results = len(results)
        if num_results == 1:
            return results[0]
        if num_results > 1:
            raise MultipleObjectsFound()
        raise ObjectNotFound()

    def to_items(self, results):
        """
        Turn LDAPResult objects returned from the ldap library into more
        convenient objects.
        """
        return [self.result_item_class(item) for item in results]

    def authenticate(self, dn='', password=''):
        """
        Attempt to authenticate given dn and password using a bind operation.
        Return True if the bind is successful, and return False there was an
        exception raised that is contained in
        self.failed_authentication_exceptions.
        """
        try:
            self.connection.simple_bind_s(dn, password)
        except tuple(self.failed_authentication_exceptions):
            return False
        else:
            return True

    def compare(self, dn, attr, value):
        """
        Compare the ``attr`` of the entry ``dn`` with given ``value``.

        This is a convenience wrapper for the ldap library's ``compare``
        function that returns a boolean value instead of 1 or 0.
        """
        return self.connection.compare_s(dn, attr, value) == 1
